### DISCUSSION TITLE
Google Chart - How to use it

### DISCUSSION THREAD

**Discussion by ottelo9** [arendst/Tasmota#15063] [Q&A] [answered]
Hi!
I use Tasmota for my smart meter to read actual and used power (W and kWh). No I want to display the values as a graph directly on the tasmota web page. I added the "USE_GOOGLE_CHARTS" compiler directive and tried to unterstand the Google Chart section in the script docu. But I dont understand how to put the actual power into an array and this array into google chart.

Example is:
$<div id="chart1"style="width:640px;height:480px;margin:0 auto"></div>
$gc(c array1 array2 "wr" "pwr1" "pwr2" "mo|di|mi|do|fr|sa|so" "Solar feed")

**Comment by ottelo9** [comment]
anyone?

**Comment by gemu2015** [comment, marked as answer]
this is an example of an obis meter with google graph

[solor_einspeisung_script.txt.zip](https://github.com/arendst/Tasmota/files/8213946/solor_einspeisung_script.txt.zip)

**Reply by ottelo9** [reply]
Very important note:
Remove the comments (and the one #) inside the >W web section! Otherwise the Main page hang.
(v11)

**Reply by gemu2015** [reply]
this should have been fixed. use latest dev

**Reply by ottelo9** [reply]
@gemu2015 
I saw your sample script on creationx.de forum ^^, very nice! Actually I try to add a new chart with power (Watt) for last 4 hours, for more detail. I put a power value every 15s into my new array[size=960=>4h]. The index counts automatically, so I dont need to do it (very useful). But I really dont know how to add a nice and correct hAxis (timeline). I tried few things with "cntN/x" but this dont work. Do you have any hint for me? Actually (see pic) I use `cstrr="cnt0"`.

Next problem (was) is that the chart area for every chart differs. I want to have all charts the same size and optimal for every device (iphone). So I tried with width=device-size or auto. Doesnt work -> First chart have e.g. 400px, second 420px, third 440px (checked with chrome inspect) and I dont know why. I also tried to put all charts into a table -> doesnt help. Actually my workaround is this:

double % !!!
```
$<div id="chart1" style="text-align:center;width:400px;height:100%%"></div>
$gc(lt s4h "wr" "Leistung [W]" cstrr)
$var options = {
$chartArea:{left:60,width:'83%%'},
$width:'100%%',
$legend: 'none',
$vAxis:{viewWindow:{min:0,max:5000}},
$title:'Leistungsverlauf 4h [Watt]'
$};
$gc(e)
```
![2022-03-21 21_18_57-MT175 - Main Menu](https://user-images.githubusercontent.com/33122175/159356906-29ccef2f-2bd6-46bc-9fe3-cda18465ca6a.png)

**Reply by nikito7** [reply]
What is your code for chart hours? XX:XX

**Reply by gemu2015** [reply]
sample script

[color_script.pdf](https://github.com/arendst/Tasmota/files/9641133/color_script.pdf)

**Reply by ottelo9** [reply]
You can find my script here:
[ottelo blog page](https://ottelo.jimdofree.com/stromz%C3%A4hler-auslesen/)

https://bit.ly/37xaeoo

**Reply by nikito7** [reply]
Thank you. This is so confusing 😆

**Reply by nikito7** [reply]
![Screenshot_20220926-105756_Chrome](https://user-images.githubusercontent.com/45373783/192250542-92d5397b-8597-40ce-ab82-5e0e6b6efa35.jpg)

Esp8266 on the limit ⚡️

**Reply by gemu2015** [reply]
esp8266 can handle google charts without any problems. There is something else wrong

**Reply by nikito7** [reply]
Big modbus script. Arrays size.

**Reply by gemu2015** [reply]
post the script

**Reply by nikito7** [reply]
This is working. 😆 But no much more resources left. Current script is stable.

https://github.com/nikito7/easyhan/blob/dev/edpbox/eb3-script.js

**Reply by nikito7** [reply]
https://github.com/nikito7/edpbox/blob/dev/tasmota/esp8266/build.h

**Comment by gemu2015** [comment]
you set the arrays on change of power values, but you must set the array on change of time.
check the examples above how this must be done.

i see no problems with esp8266 and this kind of script.

**Reply by nikito7** [reply]
Ram issues only. Nothing new. All is working.

Array based in power is a small array.

Second array is based in minutes.

Arrays above 100 entries are instable.

Heap issues. Not code 😆

**Reply by gemu2015** [reply]
your arrays have about 100 elements, that's 400 bytes, no problem at all.

i have arrays with 1000 entries on esp8266 = 4k ram, no problem

check the free heap, may be another driver uses too much ram on your setup

\>W
free heap:{m}%heap%

**Reply by nikito7** [reply]
Just sml.

Small script

![Screenshot_20220926-181605_Chrome](https://user-images.githubusercontent.com/45373783/192340142-338ea182-0386-445c-ac16-a7147c311247.jpg)


Mega sml script

![Screenshot_20220926-181442_Home Assistant](https://user-images.githubusercontent.com/45373783/192340186-d5e1f0c2-48d0-42b7-83a3-c6e27fd7395a.jpg)

Maybe I  made some mistake yesterday.

I just use esp8266 because of power requirements of my meter han port.

Esp32-C3 not replace esp8266 unfortunately

**Reply by gemu2015** [reply]
i assume  everything above 10 kb is enough for heap on esp8266.

**Reply by sfromis** [reply]
10Kb would be very low, and risky for stability.

**Reply by nikito7** [reply]
It's ok if you dont use web interface.

I try keep around 14-15kb.

Without free ram you cannot edit script.

Sometimes I have to clean script and restart

script above 4k

**Reply by gemu2015** [reply]
use script editor, then you do not need to edit script on tasmota WEBUI.

**Comment by nikito7** [comment]
Why "lt" or "ct" in code?

Docs dont say anything about "t" I think

**Reply by gemu2015** [reply]
t means table only. only the google data table is generated. the table options must be given in the next lines.
this allows for very customized tables. the syntax of google table options must be checked in google docs.

in scripting docs it stated here:

very customized chart definition:
.....

i am aware that sometimes the docs are not clear enough. feel free to improve them
everybody can contribute to the docs

**Reply by nikito7** [reply]
I'm talking about this:

![Screenshot_20220928-085623_Chrome](https://user-images.githubusercontent.com/45373783/192722303-75c9a6c3-7a93-4b2d-baa6-22667d2756c2.jpg)

Not tested with dev

**Reply by ottelo9** [reply]
You can combine l with t. So you have a linechart with data array input. Go to the script Tasmota page and scroll down to google chart. Then to the example „very customize chart definition

**Comment by nikito7** [comment]
After restart my 24h chart is inverted. Anyone else?

**Reply by nikito7** [reply]
Not reversed but:

![Screenshot_20220930-175027_Chrome](https://user-images.githubusercontent.com/45373783/193323352-6dc22997-2d42-4913-8cfe-3a3ccde0ead7.jpg)

After restart:

![Screenshot_20220930-175258_Chrome](https://user-images.githubusercontent.com/45373783/193323377-c429f1f8-2f8c-411f-83aa-ff7a7daaf2c1.jpg)


I give up in permanent arrays with auto-index.

60 minutes no permanent chart it's just good for now

**Reply by nikito7** [reply]
Found the problem 

There is no index 0 here:

```
>D
m:p:ipwrd=0 24
...
>S
ipwrd[1]=500
ipwrd[2]=600
...
ipwrd[23]=700
ipwrd[24]=800
```

**Reply by nikito7** [reply]
Ok.... array[0] is the index 😆

```js
18:26:32.436 Array: ikwh[0] 3.00
18:26:32.438 Array: ikwh[1] 8337.64
18:26:32.439 Array: ikwh[2] 8338.20
18:26:32.441 Array: ikwh[3] 8338.67
18:26:32.443 Array: ikwh[4] 0.00
```

**Reply by nikito7** [reply]
Again... index reset to zero after restart 🤷‍♂️

Feature or bug?

**Reply by gemu2015** [reply]
array index is  not permanent. you may use a permanent variable to set it.
the array index in google charts is the starting point where to start plotting the array. (it shifts the plot NOT the array itself)
 i realized that this fact  is missing in the google chart script docs.

**Comment by nikito7** [comment]
For chart use. What is the difference between m:p:array and M:p:array ?

**Reply by sfromis** [reply]
The [docs explains](https://tasmota.github.io/docs/Scripting-Language/#script-sections):
> `m:vname`
> specifies a median filter variable with 5 entries (for elimination of outliers)
> `M:vname`
> specifies a moving average filter variable with 8 entries (for smoothing data)
> ...
> `m:p:vname` defines a permanent array.

**Reply by nikito7** [reply]
But we use big arrays... still a median filter?

```
m:ipwrm=0 60
m:epwrm=0 60
```

**Reply by gemu2015** [reply]
yes, but median is only calculated if you assign the filter to a variable without an index
var=array

**Comment by jkwim** [comment]
I am trying Google Chart for the first time and need some help.

Compiled Tasmota 14.1.0.3 with:
```

        #ifndef USE_GOOGLE_CHARTS
          #define USE_GOOGLE_CHARTS
        #endif

```
I have my data in an array:
`m:den=0 8`

```
>W
<br>
<br>
; Define window for graph 1
$<div id="chart1" style="width:640px;height:300px;margin:0 auto"></div>

$gc(l den "wr" "DE" "0.75|1.5|2.25|3.0|3.75|4.5|5.25|6")

$var options = {
$vAxis:{viewWindow:{min:0,max:5000},title: 'Energy'},
$hAxis: {title: 'Distance'},
$title:'Distance Gate Energy Levels'
$};
$gc(e)

```
This is what seen on WebGUI

![image](https://github.com/user-attachments/assets/6d6d0511-baf3-4a5e-89e6-1d9d9b9242aa)

The Google Chart directives does not seem to get interpreted.

What am I doing wrong

**Reply by jkwim** [reply]
Update:
Identified the main issue. There was an issue with my compilation as I have updated the in the wrong place.

Now it draws the graph:

```
>W
; Define window for graph 1
$<div id="chart1" style="width:640px;height:300px;margin:0 auto"></div>

$gc(lt den "wr" "DEnergy" "0.75|1.5|2.25|3.0|3.75|4.5|5.25|6.0")

$var options = {
$vAxis:{viewWindow:{min:0,max:100},title: 'Energy \%'},
$hAxis: {title: 'Distance'},
$title:'Distance Gate Energy Levels'
$};

$gc(e)

```
![image](https://github.com/user-attachments/assets/0df23009-eb17-48a7-8353-d3e14a33bf93)


I have thee questions:

1. What is the meaning of the string `"wr"` in the following line? (Since I copied the code from examples I could not find the reference to this string)
`$gc(lt den "wr" "DEnergy" "0.75|1.5|2.25|3.0|3.75|4.5|5.25|6.0")`

2. Why only the first 5 labels out of` "0.75|1.5|2.25|3.0|3.75|4.5|5.25|6.0"` are getting shown in the graph?
When I checked the page source, I see empty strings being populated for these in the data table:
```
<script language='JavaScript'>

function drawChart(){

  var cssc={'headerRow':'hRow','rowNumberCell':'hCol','tableCell':'tCell'};

  var data=google.visualization.arrayToDataTable([
['wr','DEnergy'],
['0.75',100.00],
['1.5',100.00],
['2.25',24.00],
['3.0',8.00],
['3',11.00],
['',5.00],
['',100.00],
['',100.00]]
);

  var options = {vAxis:{viewWindow:{min:0,max:100},title: 'Energy'},hAxis: {title: 'Distance'},title:'Distance Gate Energy Levels'};

  var chart=new google.visualization.LineChart(document.getElementById('chart1'));
  chart.draw(data,options);
} 

google.charts.setOnLoadCallback(drawChart); 

</script>

```

3. Is there way to auto refresh the graph?

**Reply by sfromis** [reply]
If you check the [docs](https://tasmota.github.io/docs/Scripting-Language/) after the header "Google Charts:", it starts with a description of the "gc" function, saying that your "wr" is a name, described as "name of chart". And your page source clearly shows it being used in the first table row as a header for the column with the labels, naming what the labels mean.

**Reply by jkwim** [reply]
> If you check the [docs](https://tasmota.github.io/docs/Scripting-Language/) after the header "Google Charts:", it starts with a description of the "gc" function, saying that your "wr" is a name, described as "name of chart". And your page source clearly shows it being used in the first table row as a header for the column with the labels, naming the what the labels mean.

The documentation needs improvement here. It is not really the name of the chart. It is really the First Column Header.

I added a second array with following modification:

`$gc(lt den sen "Distance" "MEnergy" "SEnergy" "0.75|1.5|2.25|3.0|3.75|4.5|5.25|6.0")`

Here is the data table now.

```
        var data = google.visualization.arrayToDataTable([
          ['Distance', 'MEnergy', 'SEnergy'],
          ['0.75', 100.00, 9.00],
          ['1.5', 100.00, 0.00],
          ['2.25', 25.00, 0.00],
          ['3.0', 4.00, 100.00],
          ['3.75', 5.00, 71.00],
          ['4.5', 5.00, 56.00],
          ['5.25', 100.00, 9.00],
          ['6.0', 100.00, 9.00]
        ]);

```

**Reply by jkwim** [reply]
![image](https://github.com/user-attachments/assets/68c38ca8-6231-43bf-8e35-2b27a857cb3c)

Moving Target Energy & Static Target Energy

**Comment by gemu2015** [comment]
2. your x axis string is too long for the default script string size (20 chars)
expand string size by >D size  e.g.  >D 40
3. add this line: $<script>setTimeout(function () { location.reload();}, 3000);</script>

**Reply by jkwim** [reply]
> 2\. your x axis string is too long for the default script string size (20 chars)
>         expand string size by >D size  e.g.  >D 40

This took care of it. Thanks.

**Reply by jkwim** [reply]
> 3\. add this line: $<script>setTimeout(function () { location.reload();}, 3000);</script>

This refreshes the whole page. I was wondering whether there is a way to refresh the data table only.

**Reply by jkwim** [reply]
> This refreshes the whole page. I was wondering whether there is a way to refresh the data table only.

Found a code snippet at https://stackoverflow.com/questions/58890411/auto-refresh-google-chart
```
  google.setOnLoadCallback(() => {
    // draw the first time
    drawChart();
    // start automatic refresh
    setTimeout(() => {
      if (!isDrawing) {
        drawChart();
      }
    }, 5000);
  });
```

**Comment by jkwim** [comment]
New Question:
Does scripting support decoding this number of levels from RESULT JSON message

`mse[1]=LD2410#Moving#Sensitivity[1]`

`00:24:54.255 MQT: stat/tasmota/RESULT = {"LD2410":{"Duration":5,"Moving":{"Gates":8,"Sensitivity":[50,50,40,30,20,15,15,15,15]},"Static":{"Gates":8,"Sensitivity":[0,0,40,40,30,30,20,20,20]}}}`

**Reply by gemu2015** [reply]
this may work:
#define USE_SCRIPT_FULL_JSON_PARSER

**Reply by jkwim** [reply]
Great! That helped.
Now I have two graphs with two data sets which shows the LD2410's Moving Human Target threshold values and actual detection values and the same info for Stationary (Static)  Human Target.

Ability to view these settings in real time will help to change the Sensitivity Thresholds and configure the sensor to more precisely make the detection. This is how the manufacturer explains it. Need to put into practical and determine how effective it would be. The triggering will happen when actual level crosses the threshold level.

If there is a way to auto refresh the graphs say once a second that would be very useful as well specially when fine tuning moving target detection.

```
$<div id="chart1" style="width:640px;height:300px;margin:0 auto"></div>
$gc(lt men mse "Distance" "MEnergy" "MThresh" "0.75|1.5|2.25|3.0|3.75|4.5|5.25|6.0")
$var options = {
$vAxis:{viewWindow:{min:0,max:100},title: 'Energy \%'},
$hAxis: {title: 'Distance'},
$title:'Moving Human Energy Levels at Distance Gate'
$};
$gc(e)
$<br>
$<div id="chart2" style="width:640px;height:300px;margin:0 auto"></div>
$gc(lt sen sse "Distance" "SEnergy" "SThresh" "0.75|1.5|2.25|3.0|3.75|4.5|5.25|6.0")
$var options = {
$vAxis:{viewWindow:{min:0,max:100},title: 'Energy \%'},
$hAxis: {title: 'Distance'},
$title:'Static Human Energy Levels at Distance Gate'
$};
$gc(e)
```

![image](https://github.com/user-attachments/assets/0096fa5c-f310-4110-9241-b95925c7b71a)

![image](https://github.com/user-attachments/assets/a5a37f91-35fc-4afd-8c0f-cc0f09513cb1)

I will post more info about the modifications that I have done to the code in the following threads as I piggy backed on this thread to get clarity on implementing Google Charts:

https://github.com/arendst/Tasmota/discussions/17555#discussioncomment-10528429
https://github.com/arendst/Tasmota/discussions/22097

**Reply by gemu2015** [reply]
@jkwim 
look at this:
https://github.com/arendst/Tasmota/discussions/15767

**Reply by jkwim** [reply]
> Found a code snippet at https://stackoverflow.com/questions/58890411/auto-refresh-google-chart
> 
> ```
>   google.setOnLoadCallback(() => {
>     // draw the first time
>     drawChart();
>     // start automatic refresh
>     setTimeout(() => {
>       if (!isDrawing) {
>         drawChart();
>       }
>     }, 5000);
>   });
> ```

Any possibility to add capability to insert this via the `gc`  construct?

**Reply by gemu2015** [reply]
since the array must be updated from script at every redraw you must use a mechanism like in the oscilloscope chart from my example above https://github.com/arendst/Tasmota/discussions/15767

**Comment by jkwim** [comment]
I tried cleaning up the script and wanted to shrink the following type of code using a subroutine.

```
men[1]=LD2410#MEnergy[1]
men[2]=LD2410#MEnergy[2]
men[3]=LD2410#MEnergy[3]
men[4]=LD2410#MEnergy[4]
men[5]=LD2410#MEnergy[5]
men[6]=LD2410#MEnergy[6]
men[7]=LD2410#MEnergy[7]
men[8]=LD2410#MEnergy[8]

```

```
#me(tvar){
 men[%tvar%]=LD2410#MEnergy[%tvar%]
}
```

and calling it

`=#me(1)`

Get the error:
`14:46:33.301 var not found: LD2410#MEnergy[%tvar%]`

Is there a way to grab these kind of data without writing individual lines? Perhaps a loop?

**Reply by gemu2015** [reply]
%var% is not valid here
anyhow the parser currently does not support script variables in this json fetch

**Reply by gemu2015** [reply]
@jkwim 
i modified scripter to support this kind of json parsing
try this:
```
for cnt 1 8 1
men[cnt]=LD2410#MEnergy[%cnt%]
next
```


[xdrv_10_scripter.ino.zip](https://github.com/user-attachments/files/16923018/xdrv_10_scripter.ino.zip)

**Reply by jkwim** [reply]
Thanks. My Gitpod repository is behind some updates.

So would it suffice if I manually pick only the following changes?

![image](https://github.com/user-attachments/assets/cfa71e3a-429c-403e-a8be-60345d76a701)

**Reply by gemu2015** [reply]
YES

**Reply by jkwim** [reply]
Not successful:

```
for cnt 1 8 1
 ;print E: men[%0cnt%] %men[cnt]%
 men[cnt]=LD2410#MEnergy[%cnt%]
 print E: men[%0cnt%] %men[cnt]%
next
```

```
00:28:13.240 MQT: tele/tasmota/SENSOR = {"Time":"2024-09-09T00:28:13","Switch1":"ON","LD2410":{"Distance":[75.0,81.0,54.0],"Energy":[79,100], "MEnergy":[29,79,31,3,2,6,3,5], "SEnergy":[4,0,0,100,94,55,37,25]}}
00:28:13.245 E:
00:28:13.250 E: Energy1 100.00
00:28:13.252 E: Energy2 0.00
00:28:13.265 E: men[1] 0.00
00:28:13.268 E: men[2] 0.00
00:28:13.270 E: men[3] 0.00
00:28:13.273 E: men[4] 0.00
00:28:13.275 E: men[5] 0.00
00:28:13.278 E: men[6] 0.00
00:28:13.280 E: men[7] 0.00
00:28:13.283 E: men[8] 0.00

```

The error message is gone which means it accepts the format.

**Reply by gemu2015** [reply]
ok, fixed

```
#if 0
      uint32_t res = JsonParsePath(gv->jo, vname, '#', &fv, str_value, sizeof(str_value));
#else
      char vname_buff[64];
      Replace_Cmd_Vars(lp, sizeof(vname), vname_buff , sizeof(vname_buff));
      uint32_t res = JsonParsePath(gv->jo, vname_buff, '#', &fv, str_value, sizeof(str_value));
#endif
```

```
for cnt 1 8 1
men[cnt]=LD2410#MEnergy[%0(cnt-1)%]
next
```

**Reply by jkwim** [reply]
Works like a charm! 
Thanks. It really did save lot of script space so that I have room for more lines now.

```
for cnt 1 8 1
  men[cnt]=LD2410#MEnergy[%0(cnt-1)%]
  mse[cnt]=LD2410#Moving#Sensitivity[%0(cnt-1)%]
  sen[cnt]=LD2410#SEnergy[%0(cnt-1)%]
  sse[cnt]=LD2410#Static#Sensitivity[%0(cnt-1)%]

  tvar=cnt*0.75
  print E: %tvar%m men[%0cnt%] %men[cnt]% mse[%0cnt%] %mse[cnt]% sen[%0cnt%] %sen[cnt]% sse[%0cnt%] %sse[cnt]%
next

```

```
16:06:57.542 CMD: LD2410MovingSens
16:06:57.554 MQT: stat/tasmota/RESULT = {"LD2410":{"Duration":5,"Moving":{"Gates":8,"Sensitivity":[50,50,40,30,20,15,15,15,15]},"Static":{"Gates":8,"Sensitivity":[0,0,40,40,30,30,20,20,20]}}}
16:06:58.246 MQT: tele/tasmota/SENSOR = {"Time":"2024-09-09T16:06:58","Switch1":"ON","LD2410":{"Distance":[69.0,51.0,40.0],"Energy":[100,100], "MEnergy":[59,100,17,17,10,2,1,2], "SEnergy":[6,0,0,100,100,100,40,24]}}
16:06:58.250 E:
16:06:58.254 E: Energy1 100.00
16:06:58.256 E: Energy2 100.00
16:06:58.265 E: 0.75m men[1] 59.00 mse[1] 50.00 sen[1] 6.00 sse[1] 0.00
16:06:58.272 E: 1.50m men[2] 100.00 mse[2] 50.00 sen[2] 0.00 sse[2] 0.00
16:06:58.280 E: 2.25m men[3] 17.00 mse[3] 40.00 sen[3] 0.00 sse[3] 40.00
16:06:58.287 E: 3.00m men[4] 17.00 mse[4] 30.00 sen[4] 100.00 sse[4] 40.00
16:06:58.295 E: 3.75m men[5] 10.00 mse[5] 20.00 sen[5] 100.00 sse[5] 30.00
16:06:58.302 E: 4.50m men[6] 2.00 mse[6] 15.00 sen[6] 100.00 sse[6] 30.00
16:06:58.311 E: 5.25m men[7] 1.00 mse[7] 15.00 sen[7] 40.00 sse[7] 20.00
16:06:58.318 E: 6.00m men[8] 2.00 mse[8] 15.00 sen[8] 24.00 sse[8] 20.00

```

**Comment by jkwim** [comment]
I notice an issue.
The following line assigns value `2.0` to Switch. Previously I got the value of `1.0` correctly.

`tele/tasmota/SENSOR = {"Time":"2024-09-09T16:06:58","Switch1":"ON", ...
`
```
>E
tvar=Switch1
print E: Switch %tvar%
```

```
17:15:52.144 MQT: tele/tasmota/SENSOR = {"Time":"2024-09-09T17:15:52","Switch1":"ON","LD2410":{"Distance":[144.0,136.0,114.0],"Energy":[69,28], "MEnergy":[5,42,52,20,5,6,7,3], "SEnergy":[5,0,0,32,23,20,16,19]}}
17:15:52.149 E:
17:15:52.151 E: Switch 2.00
```

What could be the issue?

**Reply by jkwim** [reply]
I defined a new variable called `tvar2 `and used for the purpose since `tvar `was used in another loop. The outcome was same. Value of `1 `gets updated as `2`.

**Reply by gemu2015** [reply]
yes this is a bug, here is the fix:

```
#if 0
      uint32_t res = JsonParsePath(gv->jo, vname, '#', &fv, str_value, sizeof(str_value));
#else
      // replace vars allows for looped queries
      char vname_buff[64];
      Replace_Cmd_Vars(lp, sizeof(vname), vname_buff , sizeof(vname_buff));
      uint32_t res = JsonParsePath(gv->jo, vname_buff, '#', &fv, str_value, sizeof(str_value));
#endif
      if (!res) {
        goto chknext;
      }
      if (res == 1) {
        // numeric
nexit:
        if (fp) *fp = fv;
        *vtype = NUM_RES;
        tind->bits.constant = 1;
        tind->bits.is_string = 0;
        return lp + len;
      } else {
        // string
        if (!strncmp_XP(str_value, XPSTR("ON"), 2)) {
          fv = 1;
          goto nexit;
        } else if (!strncmp_XP(str_value, XPSTR("OFF"), 3)) {
          fv = 0;
          goto nexit;
        } else {
          *vtype = STR_RES;
          tind->bits.constant = 1;
          tind->bits.is_string = 1;
          if (sp) strlcpy(sp, str_value, SCRIPT_MAX_SBSIZE);
          return lp + len;
        }
      }
```

**Reply by jkwim** [reply]
Works correctly now. Thanks!

```
04:04:46.253 MQT: tele/tasmota/SENSOR = {"Time":"2024-09-10T04:04:46","Switch1":"ON","LD2410":{"Distance":[58.0,75.0,62.0],"Energy":[84,100], "MEnergy":[61,84,23,5,5,4,1,1], "SEnergy":[1,0,0,100,93,44,45,44]}}
04:04:46.260 E: Switch 1.00
```
